Add ERC: Crosschain Token Interface (#692)
ethereum/ERCs: ERCS/erc-7802.md

pragma solidity 0.8.25;

import "@openzeppelin/contracts/token/ERC20/ERC20.sol";

/// @title IERC7802
/// @notice Defines the interface for crosschain ERC20 transfers.
interface IERC7802 is IERC165 {
    /// @notice Emitted when a crosschain transfer mints tokens.
    /// @param to       Address of the account tokens are being minted for.
    /// @param amount   Amount of tokens minted.
    event CrosschainMint(address indexed to, uint256 amount);

    /// @notice Emitted when a crosschain transfer burns tokens.
    /// @param from     Address of the account tokens are being burned from.
    /// @param amount   Amount of tokens burned.
    event CrosschainBurn(address indexed from, uint256 amount);

    /// @notice Mint tokens through a crosschain transfer.
    /// @param _to     Address to mint tokens to.
    /// @param _amount Amount of tokens to mint.
    function crosschainMint(address _to, uint256 _amount) external;

    /// @notice Burn tokens through a crosschain transfer.
    /// @param _from   Address to burn tokens from.
    /// @param _amount Amount of tokens to burn.
    function crosschainBurn(address _from, uint256 _amount) external;
}

> abstract contract CrosschainERC20 is ERC20, IERC7802 {

Author: Just removed it. Also, made the constructor a bit more generic.

The interface identifier for `IERC7802` is **`0x33331994`**, calculated according to [ERC-165](./eip-165.md) as the XOR of the function selectors of the two functions in the interface:

```solidity
bytes4 constant INTERFACE_ID_IERC7802 = 
        bytes4(keccak256("crosschainMint(address,uint256)")) ^
        bytes4(keccak256("crosschainBurn(address,uint256)"));
```

or via Solidity as 

```solidity
type(IERC7802).interfaceId
```

Implementors MUST ensure that the `supportsInterface` method of ERC-165 returns true for this interface ID to indicate support for `IERC7802`.

### Methods

**`crosschainMint`**

Mints `_amount` of token to address `_account`. 

This function works as the minting entry point for bridge contracts. Each implementation is responsible for its access control logic.

```solidity
function crosschainMint(address _account, uint256 _amount) external;
```

**`crosschainBurn`**

Burns `_amount` of token from address `_account`.

This function works as the burning entry point for bridge contracts. Each implementation is responsible for its access control logic.

```solidity
function crosschainBurn(address _account, uint256 _amount) external;
```

> ### Events

Reviewer: Adding source/destination chain id & a unique message id to events would make crosschain data more easily usable. This ~may be a hard requirement for analytics, but hard to know without seeing further build out. This is necessary for joining messages across chains, to then see inflows/outflows by chain and routes.

This gap is a main reason why you don't see this data for 3P bridges. Most sites do ~1-3 L2<>L2 bridges, then give up due to the high overhead.

Reviewer 2: +1

Author: Hi @MSilb7 !

Thank you for your suggestion. We have discussed your idea internally and agree that it has value. However, we couldn't find a minimalistic set of parameters to add to the event without making the current standard more opinionated:

- **Identifier**: Incorporating an identifier would require defining an encoding standard within the ERC, which we believe is beyond the scope of this proposal.
- **Adding `from/to` and `chainIdFrom/chainIdTo`**: Including these parameters would require changing the inputs of `crosschainMint` and `crosschainBurn` to allow the bridge to pass them. This approach is opinionated because bridges might have multiple destinations or origins, and imposing single input parameters could limit their flexibility. Moreover, event-collisions would be possible.

We see significant value in adding `msg.sender` to the `CrosschainMint` and `CrosschainBurn` events. This addition would allow integrators to identify which bridge interacts with the tokens, and does not affect the interface's neutrality in anyway. 

I believe your idea has the potential to become a standard for bridge events.

Reviewer: On `msg.sender`:
- Is the use case of `msg.sender` that you would only look at transactions sent from that address on other chains? Or only function calls (internal txs) initiating from that address?
- Is this already derivable from the transaction / internal transaction? If so then the only benefit is just that you don't need to use traces, you can just use logs.

My overall opinion is like: Without any way to clearly identify how you identify the source burn for any destination mint, it's not useful. The only value it fulfills now is like: I know that this is intended to be a cross-chain transfer, versus a same-chain transfer. But it still gives me no information about where the transfer came from or goes to. Which then opens up other questions, like "How do applications know if the burn/mint is valid?"

Reviewer 3: adding msg.sender to the event would make indexing significantly easier as we can now find all mints/burns associated with a specific bridge.

Author: We just added `msg.sender`. It was very easy to implement and can help off-chain agents.

---
eip: 7802

> title: Crosschain Token Interface

Reviewer 4: "Interface" in an ERC title doesn't add a ton of information. Most ERCs contain some kind of interface.

How about:

```suggestion
title: Permissioned Token Mint/Burn Across Chains
```

Author: Agree with the suggestion of removing "interface" from the title. I'm not convinced about using "Permissioned", as that is not inherent to the interface (it could be permissionless). What do you think about something like  "Token with Mint/Burn access Across Chains"?

> description: Token interface that allows authorized contracts to mint and burn token representations during cross-chain transfers

Reviewer 4: Similarly, you can remove some nonspecific bits from your description (and hopefully expand it further):

```suggestion
description: Allow authorized contracts to mint and burn token representations during cross-chain transfers
```

author: skeletor (@skeletor-spaceman), parti (@0xParticle), joxes (@Joxess), ng (@0xng), agus duha (@agusduha), disco (@0xDiscotech), gotzen <gotzen@defi.sucks>, 0age <0age@uniswap.org>, Mark Tyneway <mark@oplabs.co>, Zain Bacchus <zain@oplabs.co>, Matt Solomon <msolomon@oplabs.co>, Maurelian <maurelian@protonmail.ch>, Blaine Malone (@blmalone)
discussions-to: https://ethereum-magicians.org/t/erc-7802-crosschain-token-interface/21508
status: Draft
type: Standards Track
category: ERC
created: 2024-10-30
requires: 165, 5679
---

## Abstract

This standard introduces a minimal interface, `IERC7802`, for tokens to enable standardized cross-chain communication. The interface consists of two functions, `crosschainMint` and `crosschainBurn`, which allow authorized bridge contracts to mint and burn token representations during cross-chain transfers. These functions serve as the entry points for bridge logic, enabling consistent handling of token supply across chains.

> The interface also defines two standardized events, `CrosschainMint` and `CrosschainBurn`, which emit metadata, including the target address, token amount, and caller. These events facilitate deterministic indexing and monitoring of cross-chain activities by off-chain agents, such as indexers, analytics tools, and auditors.

Reviewer 4: Please choose either cross-chain or crosschain and use it consistently throughout the document.

> `IERC7802` is intentionally lightweight, ensuring minimal overhead for implementation. Its modular design enables extensibility, allowing additional features—such as mint/burn limits, transfer fees, or bridge-specific access control mechanisms—to be layered on top without modifying the base interface. 

Reviewer 4: This paragraph probably belongs (with some rewording) in the Rationale section, since it explains why this standard is designed the way it is.

## Motivation

> All rollups and multiple important sidechains (Gnosis and Polygon PoS, among others) implement canonical bridges that embed their security into some part of the network's core architecture. These bridges do not have mint/burn rights over original tokens, so they usually lock (unlock) liquidity on the native chain and then mint (burn) a non-equivalent representation on the other. Mint/burn is used because the native token is non-existent on that side, so they must create a new representation. However, each bridge implements a different interface for minting/burning on non-native chains.

Reviewer 4: > Gnosis and Polygon PoS

I'd prefer if you didn't mention specific products. It unfairly advertises them over new projects that might appear after this proposal goes Final. This isn't really a hard rule though.

The proposed interface includes the most relevant and minimal set of actions used by most of these standards. These actions also do not require any governance or owner participation, in contrast, for instance, to set rate limits.

## Specification

This ERC introduces the `IERC7802` interface.

### Interface Identification

The interface identifier for `IERC7802` is **`0x33331994`**, calculated according to [ERC-165](./eip-165.md) as the XOR of the function selectors of the two functions in the interface:

```solidity
bytes4 constant INTERFACE_ID_IERC7802 = 
        bytes4(keccak256("crosschainMint(address,uint256)")) ^
        bytes4(keccak256("crosschainBurn(address,uint256)"));
```

or via Solidity as 

```solidity
type(IERC7802).interfaceId
```

Implementors MUST ensure that the `supportsInterface` method of ERC-165 returns true for this interface ID to indicate support for `IERC7802`.

### Methods

**`crosschainMint`**

Mints `_amount` of token to address `_account`. 

> This function works as the minting entry point for bridge contracts. Each implementation is responsible for its access control logic.

Reviewer 4: You can move the access control notes on each method to the security considerations section.

---
eip: 7802
title: Crosschain Token Interface

> description: Minimal token interface for cross-chain transfers

Reviewer 4: > Minimal

Too many standards claim to be minimal. I'd recommend removing this, and using your description to further elaborate on the ideas introduced in your title.

Author: Modified this section, let me know your thoughts now

Reviewer 5: Can someone here explain how to view/claim tokens or prizes that were claimed too late because I didn't understand? I hope for the information

author: skeletor (@skeletor-spaceman), parti (@0xParticle), joxes (@Joxess), ng (@0xng), agus duha (@agusduha), disco (@0xDiscotech), gotzen <gotzen@defi.sucks>, 0age <0age@uniswap.org>, Mark Tyneway <mark@oplabs.co>, Zain Bacchus <zain@oplabs.co>, Matt Solomon <msolomon@oplabs.co>, Maurelian <maurelian@protonmail.ch>, Blaine Malone (@blmalone)
discussions-to: https://ethereum-magicians.org/t/erc-7802-crosschain-token-interface/21508
status: Draft
type: Standards Track
category: ERC
created: 2024-10-30

> requires: 165

Reviewer 4: ```suggestion
requires: 165, 5679
```

---

## Abstract

> This standard introduces a minimal interface for tokens to communicate cross-chain. It allows bridges with mint and burn rights to send and relay token transfers with a standardized API.

Reviewer 4: Not bad, but I'd like to see a bit more technical meat here. Could you sketch out _how_ your proposal operates, in addition to the description you already have.

Author: Added more technical details.

## Motivation

All rollups and multiple important sidechains (Gnosis and Polygon PoS, among others) implement canonical bridges that embed their security into some part of the network's core architecture. These bridges do not have mint/burn rights over original tokens, so they usually lock (unlock) liquidity on the native chain and then mint (burn) a non-equivalent representation on the other. Mint/burn is used because the native token is non-existent on that side, so they must create a new representation.

However, each bridge implements a different interface for minting/burning on non-native chains. For example, L1 bridged tokens need to support the `OptimismMintableERC20` specification when bridged (minted) to the OP Stack, but the `IArbToken` interface when bridged to the Arbitrum Stack.

This interface fragmentation is a massive issue in cross-chain communication among chains via third-party bridges or future canonical solutions. At this point, it is clear that every bridge would benefit from a standardized interface for minted/burnt tokens. 

> There have been different attempts in the past to standardize token-bridging interfaces. However, third-party providers are also developing cross-chain token frameworks, such as NTT by Wormhole and OFT by LayerZero. Each framework defines its features, like rate limits and fee switches, and implements its mint and burn versions. The resultant interfaces become highly specific, lacking naming conventions and structures.

Reviewer 4: ```suggestion
There have been different attempts in the past to standardize token-bridging interfaces. However, third-party providers are also developing cross-chain token frameworks. Each framework defines its features, like rate limits and fee switches, and implements its mint and burn versions. The resultant interfaces become highly specific, lacking naming conventions and structures.
```

Avoid naming specific products or vendors.

```solidity
bytes4 constant INTERFACE_ID_IERC7802 = 
        bytes4(keccak256("crosschainMint(address,uint256)")) ^
        bytes4(keccak256("crosschainBurn(address,uint256)"));
```

or via Solidity as 

```solidity
type(IERC7802).interfaceId
```

Implementors MUST ensure that the `supportsInterface` method of ERC-165 returns true for this interface ID to indicate support for `IERC7802`.

### Methods

**`crosschainMint`**

Mints `_amount` of token to address `_account`. 

This function works as the minting entry point for bridge contracts. Each implementation is responsible for its access control logic.

```solidity
function crosschainMint(address _account, uint256 _amount) external;
```

**`crosschainBurn`**

Burns `_amount` of token from address `_account`.

This function works as the burning entry point for bridge contracts. Each implementation is responsible for its access control logic.

```solidity
function crosschainBurn(address _account, uint256 _amount) external;
```

### Events

**`CrosschainMint`**

> MUST trigger when `crosschainMint` is successfully called. 

Reviewer 4: This leaves a lot unspecified (which is fine if it's intentional.) For example, should other non-standard functions that also effectively perform a crosschain mint also trigger this event, or is it specifically tied to this function?

Author: Yes, this is intentionally unspecified. The `CrosschainMint` event MUST be triggered when the `crosschainMint` function is successfully called. However, implementers are not restricted from emitting this event when using other non-standard functions that effectively perform a crosschain mint.

If the implementation adheres to this standard, the full interface—including `crosschainMint` and `crosschainBurn`—will already be available, so it is expected that these functions are used for crosschain minting and burning.  That said, the event is not mandated for every action that triggers a crosschain operation, as this standard focuses specifically on the standardized interface.

The `_sender` parameter MUST be set to the msg.sender at the time the function is called.

```solidity
event CrosschainMint(address indexed _to, uint256 _amount, address indexed _sender);
```

> Note: implementations might consider additionally emitting `Transfer(address(0), _to, _amount)` to be compliant with [ERC-5679](./eip-5679.md).

Reviewer 6: I _strongly_ suggest making the Transfer events mandatory. 

ERC-20 is more relevant than ERC-5679 IMO, and in the former it's already specified as SHOULD (i.e., a strong recommendation):
>A token contract which creates new tokens SHOULD trigger a Transfer event with the _from address set to 0x0 when tokens are created.

Author: Very good catch. Just pushed to modify this invariant.

- Agents that mint and burn tokens in cross-chain transfer fundamentally differ from token owners. It make sense for the two actors to have different permissions.

Therefore, it is reasonable to have different checks, access controls, and logic (such as mint/burn limits) for cross-chain actions.

**Separation of concerns**

Merging local and cross-chain minting/burning into the same functions can lead to complex implementations that intertwine different operational logic. 
By splitting into two, concerns remain separate, making the codebase cleaner and more maintainable.

This separation of concerns is particularly relevant for

- Upgrades: Any changes in access control, limits, or logic will only affect the separate cross-chain functions (`crosschainMint` and `crosschainBurn`) without altering the standard local mint and burn implementations.
- Integrations with Different Chains: To make an [ERC-20](./eip-20.md) cross-chain compatible,
issuers simply need to implement the [ERC-7802](./eip-7802.md) extension with the corresponding access controls for each chain. 

> For example, when integrating with Optimism, the ERC-20 would grant access to the Optimism bridge; when integrating with Arbitrum, it would grant access to the Arbitrum bridge. 

Reviewer 4: Try to avoid specific vendors/products.

Author: Agree. Just removed.

This verification ensures that the token supports both `crosschainMint` and `crosschainBurn` functions, preventing scenarios where only one function is implemented. Such incomplete implementations could lead to issues like users burning tokens to bridge out but being unable to mint them upon return, resulting in failed cross-chain actions.

It is important to note that this check can only be performed locally on the chain where the token contract resides. There is no inherent guarantee that the token on the receiving chain also supports the `IERC7802` interface. Ensuring cross-chain consistency of interface support is the responsibility of the bridge implementation.

## Backwards Compatibility

This proposal is fully backwards compatible with [ERC-20](./eip-20.md).

As discussed in the Motivation section, a minimal, flexible cross-chain standard interface is necessary. The problem becomes larger as more tokens are deployed without a standardized format.

- Upgradable tokens can be upgraded to implement the new interface.

> - Non-upgradable tokens cannot implement the interface on the token itself. They can still migrate to a standard-compliant version using a lockbox mechanism, as proposed by xERC-20. The idea is to lock non-mintable tokens and mint the same amount of interface-compliant tokens. The bridge contract can act as a lockbox on the native chain.

Reviewer 4: Does xERC-20 have an ERC? If so, you should link it.

Bridge contracts will also need an upgrade to integrate with the interface. Most popular bridges in terms of liquidity at the date are upgradable or use tokens that can be upgraded to point to new bridge contracts. 

> 1. Arbitrum One ($5.23B)

Reviewer 4: Would be best to remove this leaderboard.

Author: yes! removed

2. Portal Wormhole ($2.43B)
3. Polygon PoS ($2.15B)
4. OP Mainnet ($2.09B)
5. Mantle ($983M)
6. LayerZero OFTs ($465M)
7. Omnibridge ($449M)
8. Base ($350M)
9. Scroll ($128M)
10. zkSync ($77M)

Source: L2Beat (10/31/24)

## Reference Implementation

```solidity

> // SPDX-License-Identifier: MIT

Reviewer 4: Code inline in the document must be CC0-1.0 licensed. If you have to keep it MIT, we do permit that in the assets directory.

Author: fixed
